### DISCUSSION TITLE
In-app updater

### DISCUSSION THREAD

**Discussion by StormProductionsMusic** [BearWare/TeamTalk5#1739] [Ideas]
### Description

Whenever there's a new update for TeamTalk, and we're getting the new update dialog, instead of going to the download page when pressing yes, download and install the new update through the client itself.

### Application

- [X] qtTeamTalk
- [ ] TeamTalkAndroid
- [ ] iTeamTalk
- [ ] TeamTalkServer

**Comment by bear101** [comment]
Alternatively use Microsoft Store where you'll also find TeamTalk.

**Comment by CoBC** [comment]
Hi,
it's probably easy to do this in Windows and MacOS where we just have to download new installer and tell Qt to run it. On Linux it seams to be really more hard because we have to download appropriate archive for the running platform, extract files, replaces existing ones, so really don't know how to do this easiliy. @bear101 do you have any idea on it?

**Reply by bear101** [reply]
No, it's also a bit tricky on macOS and Windows due to write permissions on running programs. But it can probably be done...

**Comment by CoBC** [comment]
Do you see any way to easily determine if we're running installed or portable version on Windows? If we run portable we can't just download new installer and run it :(

**Comment by peter1384** [comment]
currently the only way that you can check if the user is running portable or installed is to check if the file teamtalk5.ini is is not in the same level as teamtalk5.exe, then that means user is running an installed version

**Comment by CoBC** [comment]
And if a strange user move teamtalk5.ini in Program Files, or install TT 
to a different folder and than place teamtalk5.ini in the same folder? 
It's not a good solution for me

**Comment by valiant8086** [comment]
Hi.

1. check for TeamTalk5.ini in the top level of program directory, ensuring that it won't match TeamTalk5.ini.default because that is present in the installed copy.
2. place an extra file in the portable zip called portable or something like that, if file is present then it's assumed to be portable
3. make the exe or one of its dependancies compile differently for portable version so it can have a variable to check
4. have no zip file to download anymore but instead the installer can allow user to select the option to create portable and then browse to desired location for the folder to go. Add a command line argument to installer for path to executable, which will allow a running Team Talk client that has downloaded the installer to run it supplying its own running location, which provides ability for the installer to know where the portable version is and be able to update it just like an installed copy would. Probably also need a second command line argument that makes it install in portable mode or in installer mode, and also makes it install silently so user doesn't have to hit next and all that. Installer needs to close Team Talk via what ever means, window message, etc. then just install all the files in same location as the existing teamtalk exe, as determined by that command line argument.

The config files go in %appdata%\BearWare.dk when installing it, which is presumably created by Team Talk client itself, not the installer. So you can handle figuring out which is which during initial install by placing a TeamTalk5.ini in the root directory of install location if it will be portable, and not doing so if it was installed with desktop shortcut and such. Just teach the client itself to check for the ini in root directory, if not found, then check for it in %appdata%\BearWare.dk, and if not found in either location, create it there. The installer then just needs to know where the teamtalk exe is running from, and it can always try to close the client if already running, update all files it has and close out, when run in silent update mode. Just don't include the ini file when run in silent update mode so it won't accidentally put it in portable mode, lol.

1. when installing, a check box on the installer allows to install portably. When checked, nothing is registered - no shortcuts etc, and an empty/stock TeamTalk5.ini is created if not already existing in the root directory, where the exe is.
2. when unchecked, installer behaves as it does now, ensure that it does not create TeamTalk5.ini in the root directory
3. command line argument is implemented making installer support updating existing copy, which will include the facility for the client to inform the installer the location to update the files in. This will make installer:
        A. run silently
        B. kill/wait Team Talk Client
        C. install all files to the location supplied in the argument, which never include TeamTalk5.ini
        D. never mess with shortcuts. An update shouldn't need to change those, so that won't be something to distinguish whether it's portable or not
        E. Launch Team Talk Client
4. Team Talk client is modified to check on launch for TeamTalk5.ini in root directory first on launch, if not found there, check %appdata%\BearWare.dk\TeamTalk5.ini, create there if not found. Allows the client to more or less seamlessly handle either an installed or a portable situation but it does have the caveat that deleting TeamTalk5.ini from root directory will force it into install mode, creating a %appdata%\BearWare.dk folder and placing a TeamTalk5.ini in it even if it was not actually supposed to be installed. But that feels to me like enough of an edge case, and a user who was able to cause that to happen would likely understand it.
5. may as well set a constant in client to be able to indicate portable or install mode
6. when implementing new self-updater, it will download installer and run it with the command line switches/arguments necessary to get the installer to go into update mode and know the destination for the files and then try to immediately close itself out. Closing itself should help get the files unlocked but the installer will still know to try to close if it's still running, which could happen if multiple instances were open. It might be worth thinking up a way to close all instances actually.
